maven-core/src/main/java/org/apache/maven/lifecycle/providers/AbstractLifecycleProvider.java

```diff
@@ -0,0 +1,64 @@
+package org.apache.maven.lifecycle.providers;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+
+import javax.inject.Provider;
+
+import org.apache.maven.lifecycle.Lifecycle;
+import org.apache.maven.lifecycle.mapping.LifecyclePhase;
+
+/**
+ * Base lifecycle provider.
+ */
+public abstract class AbstractLifecycleProvider
+    implements Provider<Lifecycle>
+{
+    private final Lifecycle lifecycle;
+
+    protected AbstractLifecycleProvider( String id, String[] phases, String[] pluginBindings )
+    {
+        HashMap<String, LifecyclePhase> defaultBindings = null;
+        if ( pluginBindings != null )
+        {
+            int len = pluginBindings.length;
+            defaultBindings = new HashMap<>();
+            for ( int i = 0; i < len; i++ )
+            {
+                defaultBindings.put( pluginBindings[i++], new LifecyclePhase( pluginBindings[i] ) );
```

Comment:
As one coming from N. Wirth school, I find this problematic :smile:

Reply:
Can you explain?

Reply:
instead of copy pasting much code for instantiating classes, everything is done in Abstract from String[] values provided by each providers

read the Providers and you will see how it's easier to read (and later update documentation https://maven.apache.org/ref/3.8.4/maven-core/lifecycles.html )

Reply:
> As one coming from N. Wirth school, I find this problematic smile

if it was a serious algorithm, I would agree
given what it is = basic 1 line data manipulation, I'm pragmatic
of course, if you have a better way to write that algorithm, don't hesitate to improve

Reply:
Can we apply the same sanity check as with the previous PR?

Reply:
done in https://github.com/apache/maven/commit/1053bf4776dc24f59e7fb6e3772ad09839120400